xbmc/addons/interfaces/GUI/controls/Spin.cpp

```diff
@@ -0,0 +1,296 @@
+                          __FUNCTION__, addon, control, value, addon ? addon->ID().c_str() : "unknown");
+    return;
+  }
+
+  control->SetStringValue(std::string(value));
+}
+
+char* Interface_GUIControlSpin::get_string_value(void* kodiBase, void* handle)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  CGUISpinControlEx* control = static_cast<CGUISpinControlEx*>(handle);
+  if (!addon || !control)
+  {
+    CLog::Log(LOGERROR, "Interface_GUIControlSpin::%s - invalid handler data (kodiBase='%p', handle='%p') on addon '%s'",
+                          __FUNCTION__, addon, control, addon ? addon->ID().c_str() : "unknown");
+    return nullptr;
+  }
+
+  return strdup(control->GetStringValue().c_str());
+}
+
+void Interface_GUIControlSpin::add_int_label(void* kodiBase, void* handle, const char* label, int value)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  CGUISpinControlEx* control = static_cast<CGUISpinControlEx*>(handle);
+  if (!addon || !control || !label)
+  {
+    CLog::Log(LOGERROR, "Interface_GUIControlSpin::%s - invalid handler data (kodiBase='%p', handle='%p', label='%p') on addon '%s'",
+                          __FUNCTION__, addon, control, label, addon ? addon->ID().c_str() : "unknown");
+    return;
+  }
+
+  control->AddLabel(std::string(label), value);
+}
+
+void Interface_GUIControlSpin::set_int_range(void* kodiBase, void* handle, int start, int end)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  CGUISpinControlEx* control = static_cast<CGUISpinControlEx*>(handle);
+  if (!addon || !control)
+  {
+    CLog::Log(LOGERROR, "Interface_GUIControlSpin::%s - invalid handler data (kodiBase='%p', handle='%p') on addon '%s'",
+                          __FUNCTION__, addon, control, addon ? addon->ID().c_str() : "unknown");
+    return;
+  }
+
+  control->SetRange(start, end);
+}
+
+void Interface_GUIControlSpin::set_int_value(void* kodiBase, void* handle, int value)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  CGUISpinControlEx* control = static_cast<CGUISpinControlEx*>(handle);
+  if (!addon || !control)
+  {
+    CLog::Log(LOGERROR, "Interface_GUIControlSpin::%s - invalid handler data (kodiBase='%p', handle='%p') on addon '%s'",
+                          __FUNCTION__, addon, control, addon ? addon->ID().c_str() : "unknown");
+    return;
+  }
+
+  control->SetValue(value);
+}
+
+int Interface_GUIControlSpin::get_int_value(void* kodiBase, void* handle)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  CGUISpinControlEx* control = static_cast<CGUISpinControlEx*>(handle);
+  if (!addon || !control)
+  {
+    CLog::Log(LOGERROR, "Interface_GUIControlSpin::%s - invalid handler data (kodiBase='%p', handle='%p') on addon '%s'",
+                          __FUNCTION__, addon, control, addon ? addon->ID().c_str() : "unknown");
+    return 0;
```

Comment:
Not sure what's the best, has set it as 0 to have equal with float error return and is normally the default value if nothing becomes set before.